[FIX] web_editor, website: consider other media types in s_image_gallery

sobo-odoo opened the pull request:
WIP

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/166061).

sobo-odoo commented:
@qsm-odoo ready for review :slightly_smiling_face: 

Some remarks/explanations:
- About the `isMediaElement` function you suggested to use, I used it at first in order to get all the media of the target, but in the end I removed it for the following reasons:
  - we said we should not support icons (and I deduced documents too) as it is really unlikely for a user to add an icon in this snippet, so we only need to get the images and the videos (so `img` and `.media_iframe_video`), which is what I therefore did
  - when there was a carousel (so "Image Gallery"), it was also getting the carousel arrows so the media needed to be filtered and as it became too much, I dropped it.
 
- About what I asked for the thumbnail url, in the end I kept the route instead of doing a JS utils, because in either cases it would have been async because of vimeo => the hardcoded link leads to an object and we need to do a request on it to get the thumbnail link, which should be awaited. So there is no way to fetch the url without async code, which is why the options were modified to become async when needed :+1: 

- For the issue about the media dialog, I triggered a new `init_media_dialog` event when the media dialog is initialized, in order to get what I need from the clicked media. And I noticed that by passing the options in the event, we can modify them in the handler, allowing to set them as we want (so I only kept images and videos). Not sure if it is really correct but it's the only way I found without making the media dialog manually (like I did before) :grimacing:
Note that this event is triggered only when we replace a media (so not when adding new images) (because the trigger is on the media, so we need to have double clicked on one) :wink: 

- About the big diff ( :grimacing:) , it is mainly because there is a lot of variable renaming from `imageEls, imageHolderEls, images, imgs...` to `mediaEls, mediaHolderEls`. I know these could be considered as "cosmetic" changes but I discussed it with BSO before, and we concluded that it was well a fix, as it would allow to know that there are videos too, which would prevent doing something only image related in eventual future bug fixes   

If you don't remember any of these points, don't hesitate to ask me (I was lazy to explain in details here) :slightly_smiling_face: 

For the security flags, it is because of the imports in python, but I guess it is okay as it the same code as in 16.0 and the URL comes from the media dialog, which does not upload the video unless it is a valid URL :thinking:

qsm-odoo commented on addons/web_editor/tools.py:
```diff
@@ -138,27 +138,43 @@ def get_video_embed_code(video_url):
+        thumbnail_url = f'https://img.youtube.com/vi/{video_id}/0.jpg'
     elif platform == 'vimeo':
-        res = requests.get(f'http://vimeo.com/api/oembed.json?url={video_url}', timeout=10)
+        video_details = f'http://vimeo.com/api/oembed.json?url={video_url}'
+        res = requests.get(video_details, timeout=10)
```
Nitpicking: maybe this is an useless change :shrug:

sobo-odoo replied:
It was to make it clear that the vimeo url leads to an object containing the video details and not directly to the thumbnail, as I didn't find it straightforward to understand :slightly_smiling_face: 
Should I still remove this change ?

qsm-odoo replied:
If you think this helps, then you can keep it. Although, name it `video_details_url` or `url_to_video_details` then (this is what this variable contains: an URL) :+1:

qsm-odoo commented on addons/web_editor/tools.py:
```diff
@@ -138,27 +138,43 @@ def get_video_embed_code(video_url):
+        return None
+
+    response = None
+    if platform in ['youtube', 'dailymotion', 'instagram', 'vimeo']:
+        response = requests.get(thumbnail_url, timeout=10)
```
Is that `if` necessary?

sobo-odoo replied:
Yes, because it should not be done for "Youku" for example :slightly_smiling_face: I could have done "if not Youku", but I think it is better to include the working cases and update the list when new cases will be added, than to exclude every other ones (in the end, I only grouped all the if's) :slightly_smiling_face:

qsm-odoo replied:
Oh ok, I was confused by the fact `get_video_thumbnail_url` only handle those 4 ones. Then it means that in the case of other platforms we don't have thumbnails? Would not `if thumbnail_url:` be more straightforward here then? :thinking: That would avoid "duplicating" the list of platforms to consider and actually do what you are meant to do "if we got a thumbnail_url, use it"?

sobo-odoo replied:
Aah indeed, and in the end, the `if` is not even necessary, because the condition just above returns if there is no thumbnail...
My bad I answered too quickly :grimacing:

qsm-odoo commented on addons/website/controllers/main.py:
```diff
@@ -799,6 +800,14 @@ def actions_server(self, path_or_xml_id_or_id, **post):
+    # Video information
+    # ------------------------------------------------------
+
+    @http.route(['/website/get_video_thumbnail_url'], type='json', auth='user', website=True)
+    def get_video_thumb_url(self, video_url):
+        return get_video_thumbnail_url(video_url)
```
The `auth="user"` protection is probably good enough but maybe we can go further and check if the user has the restricted editor rights? :pray: Otherwise it means that portal users can use any Odoo instance to fetch third party video thumbnails... probably not that big of a deal but :shrug:

qsm-odoo commented on addons/website/static/src/snippets/s_image_gallery/001.scss:
```diff
@@ -78,7 +78,16 @@
+                position: relative;
                 border: 1px solid #aaa;
+
+                // Play icon for video thumnail indicators.
```
Typo :)

qsm-odoo commented on addons/website/static/src/snippets/s_image_gallery/options.js:
```diff
@@ -32,12 +32,44 @@ options.registry.gallery = options.Class.extend({
-                .css('background-image', 'url(' + $img.attr('src') + ')');
+        // Manage the media replacement. Make sure the data-index attribute and
+        // the media previews are updated if medias are changed.
+        // TODO init_media_dialog is not triggered anymore
```
I added this comment here as a reminder, not sure if there is an appropriate place in 16.0 to trigger this event. Maybe there is a better way in 16.0 who knows :shrug: I let you check :+1:

qsm-odoo commented on addons/website/static/src/snippets/s_image_gallery/options.js:
```diff
@@ -32,12 +32,44 @@ options.registry.gallery = options.Class.extend({
+            // Make sure the steps are unpaused when closing the dialog.
+            data.options.onClose = () => this.options.wysiwyg.odooEditor.historyUnpauseSteps();
+        });
+        // TODO replace_target is not triggered anymore
+        this.$target.on("replace_target.gallery", "img, .media_iframe_video", async (_, mediaEl) => {
```
I added the comment as a reminder, I don't think this event exist in 16.0. So if possible, find another way. If not possible, restore the event trigger? :shrug:

qsm-odoo commented on addons/website/static/src/snippets/s_image_gallery/options.js:
```diff
@@ -110,8 +154,9 @@ options.registry.gallery = options.Class.extend({
         var $container = this.$('> .container, > .container-fluid, > .o_container_small');
-        const lastImage = _.last(this._getImages());
-        let index = lastImage ? this._getIndex(lastImage) : -1;
+        const mediaEls = this._getImages();
+        const lastMediaEl = mediaEls[mediaEls.length - 1];
+        let index = lastMediaEl ? this._getIndex(lastMediaEl) : -1;
```
Not sure we need this change in stable :wink:

sobo-odoo replied:
But what if we replace an image by a video (considering that my fix properly keeps the `data-index` on it) and then add new images ? If the last media was the video, without this change, it will get the index of the last image and so the new images indexes will be wrong, no ? Since they will not start with the real last index, which is the video one :thinking:

qsm-odoo replied:
@sobo-odoo I must be blind but my point is that the previous code does the exact same thing as your new code (am I wrong?)

Before:
```javascript
const lastImage = _.last(this._getImages());
let index = lastImage ? this._getIndex(lastImage) : -1;
```
After:
```javascript
const mediaEls = this._getImages();
const lastMediaEl = mediaEls[mediaEls.length - 1];
let index = lastMediaEl ? this._getIndex(lastMediaEl) : -1;
```
In both cases, are not you getting the last item returned by `_getImages` and in both case calling `_getIndex` with it ?

sobo-odoo replied:
Aah my bad, I forgot it was still called `_getImages()` :smiling_face_with_tear: 

But I mentioned in my message before your review that a lot of changes were variables renaming, because it would allow to tell that there are not only images, which would avoid future issues, and were therefore not "cosmetic" changes

> About the big diff ( 😬) , it is mainly because there is a lot of variable renaming from imageEls, imageHolderEls, images, imgs... to mediaEls, mediaHolderEls. I know these could be considered as "cosmetic" changes but I discussed it with BSO before, and we concluded that it was well a fix, as it would allow to know that there are videos too, which would prevent doing something only image related in eventual future bug fixes


(it also allowed to remove jQuery code)

Does that mean that you do not agree with this type of changes ?

qsm-odoo replied:
It's ok :+1: Let's keep it :+1:

qsm-odoo commented on addons/website/static/src/snippets/s_image_gallery/options.js:
```diff
@@ -236,31 +288,36 @@ options.registry.gallery = options.Class.extend({
                     $lowest = $col;
                 }
             });
-            // Only on Chrome: appended images are sometimes invisible and not
-            // correctly loaded from cache, we use a clone of the image to force
-            // the loading.
-            $lowest.append(imgs.shift().cloneNode(true));
```
Note: during conflict resolution, there are 3-4 places where you changed a `img` into a `mediaHolderEl` but where in 16.0 a `.cloneNode(true)` was added (see here with a comment explaining why). During conflict resolution, I just ignored that and kept your version with `mediaHolderEl` but probably worth checking if any issue due to not cloning :pray:

qsm-odoo reviewed (changes requested):
@sobo-odoo Given the security validation needed + amount of change + backport of utils + non critical bug, I am choosing to not target 15.0. Let's target 16.0 where the codebase is significantly closer to the master version :+1: 

I rebased your branch on top of 16.0, retargeted the PR and also squashed both commits (given there is no "backport of utils" anymore, it's better to have the utils improvement as part of the main commit which uses those improvements).

There were many conflicts, I solved them and added comments about some of them in my review here. I did not test the branch after my conflict resolution, I am letting you do it :pray: Sorry for the delay checking, I'll try to merge this faster once you confirm it is good to go for 16.0 (see my sub-comments in this review), thank you :pray:
